meta/Decays.m

```diff
@@ -0,0 +1,2382 @@
+           _, Print["Error: cannot determine allowed generic final states for n = ", n]; Quit[1];
+          ];
+
+GetAllowedGenericFinalStates[particle_?TreeMasses`IsFermion, n_Integer] :=
+    Switch[n,
+           2, {{GenericScalarName[], GenericFermionName[]},
+               {GenericVectorName[], GenericFermionName[]}},
+           _, Print["Error: cannot determine allowed generic final states for n = ", n]; Quit[1];
+          ];
+
+GetVerticesForDecay[decay_FSParticleDecay] :=
+    Module[{diagrams = GetDecayDiagramsOnly[decay]},
+           DeleteDuplicates[Join @@ (Flatten[(CXXDiagrams`VerticesForDiagram /@ Last[#]), 1]& /@ diagrams)]
+          ];
+
+GetVerticesForDecays[particleDecays_List] :=
+    DeleteDuplicates[Flatten[GetVerticesForDecay /@ particleDecays, 1]];
+
+LoopOverIndex[loopBody_String, index_, start_, stop_, type_:CConversion`ScalarType[CConversion`integerScalarCType]] :=
+    Module[{idxStr, startStr, stopStr},
+           idxStr = CConversion`ToValidCSymbolString[index];
+           startStr = CConversion`ToValidCSymbolString[start];
+           stopStr = CConversion`ToValidCSymbolString[stop];
+           "for (" <> CConversion`CreateCType[type] <> " " <> idxStr <>
+           " = " <> startStr <> "; " <> idxStr <> " < " <> stopStr <> "; ++" <>
+           idxStr <> ") {\n" <> TextFormatting`IndentText[loopBody] <> "\n}"
+          ];
+
+(* generates a loop over the given indices, in the form
+   {{idx1, start, stop}, {idx2, start, stop}, ...} with
+   the first list entry being the innermost loop *)
+LoopOverIndexCollection[loopBody_String, indices_List] :=
+    "\n" <> Fold[LoopOverIndex[#1, Sequence @@ #2]&, loopBody, indices] <> "\n";
+
+CreateGenericGetPartialWidthFunctionName[] := "get_partial_width";
+
+CreateSpecializedPartialWidthCalculationName[initialState_, finalState_List, fieldsNamespace_] :=
+    CreateGenericGetPartialWidthFunctionName[] <> "<" <>
+    CXXDiagrams`CXXNameOfField[initialState, prefixNamespace -> fieldsNamespace] <> "," <>
+    Utils`StringJoinWithSeparator[CXXDiagrams`CXXNameOfField[#, prefixNamespace -> fieldsNamespace]& /@ finalState, ","] <> " >";
+CreateSpecializedPartialWidthCalculationName[initialState_, finalState_List] :=
+    CreateGenericGetPartialWidthFunctionName[] <> "<" <>
+        CXXDiagrams`CXXNameOfField[initialState] <> ", " <>
+        Utils`StringJoinWithSeparator[CXXDiagrams`CXXNameOfField[#]& /@ finalState, ", "] <> ">";
+
+CreatePartialWidthCalculationName[decay_FSParticleDecay, scope_:""] :=
+    Module[{initialState, initialStateName,
+            finalState, finalStateName},
+           initialState = GetInitialState[decay];
+           initialStateName = CConversion`ToValidCSymbolString[initialState];
+           finalState = GetFinalState[decay];
+           finalStateName = StringJoin[CConversion`ToValidCSymbolString /@ finalState];
+           scope <> If[scope != "", "::", ""] <> "partial_width_" <> initialStateName <> "_to_" <> finalStateName
```

Comment:
We should apply `IndentText` to the return value of this function for proper indentation. :)